File: Python/pong.py
import pygame
import sys
import random

# Constants
WINDOW_WIDTH = 640
WINDOW_HEIGHT = 480
BALL_RADIUS = 10
PADDLE_WIDTH = 15
PADDLE_HEIGHT = 60
PADDLE_SPEED = 5
BALL_SPEED_X = 3
BALL_SPEED_Y = 3

# Colors
BLACK = pygame.Color('black')
WHITE = pygame.Color('white')

# Initialize Pygame
pygame.init()

# Set up window
window = pygame.display.set_mode((WINDOW_WIDTH, WINDOW_HEIGHT))
pygame.display.set_caption('Pong')

# Set up clock
clock = pygame.time.Clock()

# Set up ball
ball = pygame.Rect(WINDOW_WIDTH // 2 - BALL_RADIUS, WINDOW_HEIGHT // 2 - BALL_RADIUS, BALL_RADIUS * 2, BALL_RADIUS * 2)
ball_speed_x = BALL_SPEED_X
ball_speed_y = BALL_SPEED_Y

# Set up paddles
paddle_a = pygame.Rect(0, WINDOW_HEIGHT // 2 - PADDLE_HEIGHT // 2, PADDLE_WIDTH, PADDLE_HEIGHT)
paddle_b = pygame.Rect(WINDOW_WIDTH - PADDLE_WIDTH, WINDOW_HEIGHT // 2 - PADDLE_HEIGHT // 2, PADDLE_WIDTH, PADDLE_HEIGHT)

# Set up fonts
font = pygame.font.Font(None, 36)

# Initialize scores
score_a = 0
score_b = 0

# Function to reset the game
def reset_game():
    global score_a, score_b
    ball.center = (WINDOW_WIDTH // 2, WINDOW_HEIGHT // 2)
    ball_speed_x = BALL_SPEED_X
    ball_speed_y = BALL_SPEED_Y
    score_a = 0
    score_b = 0

# Main game loop
while True:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            pygame.quit()
            sys.exit()

    # Update ball position
    ball.x += ball_speed_x
    ball.y += ball_speed_y

    # Check collision with paddles
    if ball.colliderect(paddle_a) or ball.colliderect(paddle_b):
        ball_speed_x *= -1

    # Check collision with top and bottom walls
    if ball.top <= 0 or ball.bottom >= WINDOW_HEIGHT:
        ball_speed_y *= -1

    # Check if ball goes off the screen
    if ball.left <= 0:
        # Player B scores
        score_b += 1
        reset_game()
    elif ball.right >= WINDOW_WIDTH:
        # Player A scores
        score_a += 1
        reset_game()

    # Update paddles
    keys = pygame.key.get_pressed()
    if keys[pygame.K_UP] and paddle_b.top > 0:
        paddle_b.y -= PADDLE_SPEED
    if keys[pygame.K_DOWN] and paddle_b.bottom < WINDOW_HEIGHT:
        paddle_b.y += PADDLE_SPEED

    # AI-controlled paddle for Player A
    # Paddle A will move towards the ball's y-coordinate
    if paddle_a.centery < ball.centery and paddle_a.bottom < WINDOW_HEIGHT:
        paddle_a.y += PADDLE_SPEED
    elif paddle_a.centery > ball.centery and paddle_a.top > 0:
        paddle_a.y -= PADDLE_SPEED

    # Clear screen
    window.fill(BLACK)

    # Draw ball
    pygame.draw.circle(window, WHITE, (ball.x + BALL_RADIUS, ball.y + BALL_RADIUS), BALL_RADIUS)

    # Draw paddles
    pygame.draw.rect(window, WHITE, paddle_a)
    pygame.draw.rect(window, WHITE, paddle_b)

    # Draw score
    score_text = font.render(f'{score_a} - {score_b}', True, WHITE)
    score_text_rect = score_text.get_rect()
    score_text_rect.center = (WINDOW_WIDTH // 2, 20)
    window.blit(score_text, score_text_rect)

    # Draw center line
    pygame.draw.aaline(window, WHITE, (WINDOW_WIDTH // 2, 0), (WINDOW_WIDTH // 2, WINDOW_HEIGHT))

    # Update display
    pygame.display.flip()

    # Limit frames per second
    clock.tick(60)

File: Kotlin/pong.kt
import android.graphics.*
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    // Constants
    private val WINDOW_WIDTH = 640
    private val WINDOW_HEIGHT = 480
    private val BALL_RADIUS = 10
    private val PADDLE_WIDTH = 15
    private val PADDLE_HEIGHT = 60
    private val PADDLE_SPEED = 5
    private val BALL_SPEED_X = 3
    private val BALL_SPEED_Y = 3

    // Colors
    private val BLACK = Color.BLACK
    private val WHITE = Color.WHITE

    // Variables
    private var ballX = 0
    private var ballY = 0
    private var ballSpeedX = BALL_SPEED_X
    private var ballSpeedY = BALL_SPEED_Y
    private var paddleAY = 0
    private var paddleBY = 0
    private var scoreA = 0
    private var scoreB = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Set up ball
        ballX = WINDOW_WIDTH / 2 - BALL_RADIUS
        ballY = WINDOW_HEIGHT / 2 - BALL_RADIUS

        // Set up paddles
        paddleAY = WINDOW_HEIGHT / 2 - PADDLE_HEIGHT / 2
        paddleBY = WINDOW_HEIGHT / 2 - PADDLE_HEIGHT / 2

        // Set up scores
        scoreA = 0
        scoreB = 0

        // Set up game loop
        val gameLoop = object : Runnable {
            override fun run() {
                update()
                draw()
                view.postDelayed(this, 16) // Adjust delay for desired frame rate
            }
        }

        // Start game loop
        val view = findViewById<View>(R.id.gameView) // Replace with your actual game view
        view.post(gameLoop)
    }

    private fun update() {
        // Update ball position
        ballX += ballSpeedX
        ballY += ballSpeedY

        // Check collision with paddles
        if (ballX < PADDLE_WIDTH && ballY + BALL_RADIUS >= paddleAY && ballY <= paddleAY + PADDLE_HEIGHT) {
            ballSpeedX *= -1
        } else if (ballX > WINDOW_WIDTH - PADDLE_WIDTH - BALL_RADIUS && ballY + BALL_RADIUS >= paddleBY && ballY <= paddleBY + PADDLE_HEIGHT) {
            ballSpeedX *= -1
        }

        // Check collision with walls
        if (ballY <= 0 || ballY >= WINDOW_HEIGHT - BALL_RADIUS) {
            ballSpeedY *= -1
        }

        // Check if ball goes off the screen
        if (ballX <= 0) {
            scoreB++
            resetGame()
        } else if (ballX >= WINDOW_WIDTH - BALL_RADIUS) {
            scoreA++
            resetGame()
        }

        // Update paddle B (AI-controlled)
        if (paddleBY + PADDLE_HEIGHT / 2 < ballY && paddleBY + PADDLE_HEIGHT < WINDOW_HEIGHT) {
            paddleBY += PADDLE_SPEED
        } else if (paddleBY + PADDLE_HEIGHT / 2 > ballY && paddleBY > 0) {
            paddleBY -= PADDLE_SPEED
        }
    }

    private fun draw() {
        // Get the canvas to draw on
        val canvas = surfaceHolder.lockCanvas()

        // Clear the canvas
        canvas.drawColor(BLACK)

        // Draw the ball
        val ballRect = Rect(ballX, ballY, ballX + BALL_RADIUS * 2, ballY + BALL_RADIUS * 2)
        canvas.drawCircle(ballRect.exactCenterX(), ballRect.exactCenterY(), BALL_RADIUS.toFloat(), paint)

        // Draw the paddles
        val paddleARect = Rect(0, paddleAY, PADDLE_WIDTH, paddleAY + PADDLE_HEIGHT)
        val paddleBRect = Rect(WINDOW_WIDTH - PADDLE_WIDTH, paddleBY, WINDOW_WIDTH, paddleBY + PADDLE_HEIGHT)
        canvas.drawRect(paddleARect, paint)
        canvas.drawRect(paddleBRect, paint)

        // Draw the score
        val scoreText = "Score: $scoreA - $scoreB"
        val scoreTextSize = paint.textSize
        val scoreTextBounds = Rect()
        paint.getTextBounds(scoreText, 0, scoreText.length, scoreTextBounds)
        val scoreTextX = (WINDOW_WIDTH - scoreTextBounds.width()) / 2
        val scoreTextY = (WINDOW_HEIGHT + scoreTextBounds.height()) / 2
        canvas.drawText(scoreText, scoreTextX.toFloat(), scoreTextY.toFloat(), paint)

        // Unlock the canvas and update the surface
        surfaceHolder.unlockCanvasAndPost(canvas)
    }

    // Helper function to reset the game
    private fun resetGame() {
        ballX = WINDOW_WIDTH / 2 - BALL_RADIUS
        ballY = WINDOW_HEIGHT / 2 - BALL_RADIUS
        ballSpeedX = BALL_SPEED_X
        ballSpeedY = BALL_SPEED_Y
    }
}
